vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1706 @@
+    }
+
+    /*
+     * Create mapping form tile_id to its location
+     */
+    void create_tile_id_to_loc() {
+        int max_height = 0;
+        for (const auto& tile : ar_.getTileList()) {
+            tile_to_loc_[tile.getName()] = location(tile.getCol() + 1, tile.getRow() + 1);
+            max_height = std::max(max_height, (int)(tile.getRow() + 1));
+            loc_to_tile_[location(tile.getCol() + 1, tile.getRow() + 1)] = tile.getName();
+        }
+        /* tile with name -1 is assosiated with constant source tile */
+        tile_to_loc_[-1] = location(1, max_height + 1);
+        loc_to_tile_[location(1, max_height + 1)] = -1;
+    }
+
+    /*
+     * Create uniq id for each FPGA Interchange node.
+     * Create mapping from wire to node id and from node id to its locations
+     * These ids are used later for site pins and pip conections (rr_edges)
+     */
+    void process_const_nodes() {
+        for (const auto& tile : ar_.getTileList()) {
+            int tile_id = tile.getName();
+            auto tile_type = ar_.getTileTypeList()[tile.getType()];
+            auto wires = tile_type.getWires();
+            for (const auto& constant : tile_type.getConstants()) {
+                int const_id = constant.getConstant() == Device::ConstantType::VCC ? 1 : 0;
+                for (const auto wire_id : constant.getWires()) {
+                    wire_to_node_[std::make_tuple(tile_id, wires[wire_id])] = const_id;
+                    node_to_locs_[const_id].insert(tile_to_loc_[tile_id]);
+                    if (wire_name_to_seg_idx_.find(str(wires[wire_id])) == wire_name_to_seg_idx_.end())
+                        wire_name_to_seg_idx_[str(wires[wire_id])] = -1;
+                    node_tile_to_segment_[std::make_tuple(const_id, tile_id)] = wire_name_to_seg_idx_[str(wires[wire_id])];
+                }
+            }
+        }
+        for (const auto& node_source : ar_.getConstants().getNodeSources()) {
+            int tile_id = node_source.getTile();
+            int wire_id = node_source.getWire();
+            int const_id = node_source.getConstant() == Device::ConstantType::VCC ? 1 : 0;
+            wire_to_node_[std::make_tuple(tile_id, wire_id)] = const_id;
+            node_to_locs_[const_id].insert(tile_to_loc_[tile_id]);
+            if (wire_name_to_seg_idx_.find(str(wire_id)) == wire_name_to_seg_idx_.end())
+                wire_name_to_seg_idx_[str(wire_id)] = -1;
+            node_tile_to_segment_[std::make_tuple(const_id, tile_id)] = wire_name_to_seg_idx_[str(wire_id)];
+        }
+        for (int i = 0; i < 2; ++i) {
+            wire_to_node_[std::make_tuple(-1, i)] = i;
+            node_to_locs_[i].insert(tile_to_loc_[-1]);
+            node_tile_to_segment_[std::make_tuple(i, -1)] = -1;
+        }
+    }
+
+    void create_uniq_node_ids() {
+        /*
+         * Process constant sources
+         */
+        process_const_nodes();
+        /*
+         * Process nodes
+         */
+        int id = 2;
```

Comment:
I believe that the starting id is because of the constant nodes. Probably it would be better to add a comment stating that

Reply:
Comment added in next commit